src/stan/variational/advi.hpp

```diff
@@ -522,7 +526,367 @@ class advi {
     return stan::services::error_codes::OK;
   }
 
-  // TODO(akucukelbir): move these things to stan math and test there
+  /**
+  * RVI Diagnostics: Calculates log importance weights
+  *
+  * @param[in] variational_obj variational family object
+  * @param[in, out] weight_vector An Eigen
+  * dynamic vector of weights, sorted in descending order
+  */
+  void lr(const Q& variational_obj, Eigen::VectorXd& weight_vector) 
+          const {
+    // Need to check the vector is empty
+    weight_vector.resize(n_posterior_samples_);
+    double log_p, log_g;
+    std::stringstream msg2;
+    // Draw posterior sample. log_g is the log normal densities.
+    for (int n = 0; n < n_posterior_samples_; ++n) {
+      variational_obj.sample_log_g(rng_, cont_params_, log_g);
+      //  log_p: Log probability in the unconstrained space
+      log_p = model_.template log_prob<false, true>(cont_params_, &msg2);
+      weight_vector(n) = log_p - log_g;
+    }
+    // sort descending order
+    std::sort(weight_vector.data(), weight_vector.data() + weight_vector.size());
+  }
+
+  /**
+  * RVI Diagnostics: Approximate parameters k and sigma for pareto-k diagnostics
+  * from a given sample vector x
+  * Zhang, Stephens (2009)
+  * 
+  * @param[in] x Vector of values in which pareto parameters will be estimated
+  * @param[in, out] k_ret Variable to return the estimated k value.
+  * @param[in, out] sigma Variable to return the estimated sigma value.
+  * @param[in] wip Boolean indicating whether to use a weakly informed prior
+  * @param[in] min_grid_pts The minimum number of grid points used in the fitting 
+  * algorithm. The actual number used is `min_grid_pts + floor(sqrt(length(x)))`
+  * @param[in] sort_x If `true` (the default), the first step in the fitting
+  * algorithm is to sort the elements of `x`. If `x` is already sorted in
+  * ascending order then `sort_x` can be set to `false` to skip the initial
+  * sorting step.
+
+  */
+  static void gpdfit(Eigen::VectorXd& x, double& k, double& sigma, 
```

Comment:
It looks like this function doesn't depend on any model stuff, so I would put it outside the model class somewhere. The same folder as the ESS/Rhat stuff would be good